File: python-version/file_transfer/png_codec.py
from __future__ import annotations

import struct
import zlib

from .errors import FileTransferError
from .format import FIXED_HEADER_SIZE, MAX_FILE_SIZE

PNG_SIGNATURE = b"\x89PNG\r\n\x1a\n"


def _error(code: str, message: str) -> FileTransferError:
    return FileTransferError(code, message)


def _chunk(kind: bytes, data: bytes) -> bytes:
    return struct.pack(">I", len(data)) + kind + data + struct.pack(">I", zlib.crc32(kind + data) & 0xFFFFFFFF)


def encode_rgb_png(width: int, height: int, rgb: bytes) -> bytes:
    if width <= 0 or height <= 0 or len(rgb) != width * height * 3:
        raise _error("INVALID_DIMENSIONS", "RGB 数据长度与图片尺寸不匹配")
    stride = width * 3
    raw = b"".join(b"\x00" + rgb[row * stride:(row + 1) * stride] for row in range(height))
    ihdr = struct.pack(">IIBBBBB", width, height, 8, 2, 0, 0, 0)
    return PNG_SIGNATURE + _chunk(b"IHDR", ihdr) + _chunk(b"IDAT", zlib.compress(raw, 9)) + _chunk(b"IEND", b"")


def _paeth(a: int, b: int, c: int) -> int:
    p = a + b - c
    pa, pb, pc = abs(p - a), abs(p - b), abs(p - c)
    return a if pa <= pb and pa <= pc else b if pb <= pc else c


def decode_png_to_rgb(png: bytes) -> tuple[int, int, bytes]:
    if not png.startswith(PNG_SIGNATURE):
        raise _error("INVALID_PNG", "输入不是有效 PNG")
    offset = len(PNG_SIGNATURE)
    ihdr = None
    idat = bytearray()
    seen_iend = False
    while offset < len(png):
        if offset + 12 > len(png):
            raise _error("INVALID_PNG", "PNG 数据被截断")
        length = struct.unpack(">I", png[offset:offset + 4])[0]
        kind = png[offset + 4:offset + 8]
        end = offset + 12 + length
        if end > len(png):
            raise _error("INVALID_PNG", "PNG chunk 长度无效")
        data = png[offset + 8:offset + 8 + length]
        expected_crc = struct.unpack(">I", png[offset + 8 + length:end])[0]
        if zlib.crc32(kind + data) & 0xFFFFFFFF != expected_crc:
            raise _error("INVALID_PNG", f"PNG {kind.decode('ascii', 'replace')} chunk CRC 错误")
        if kind == b"IHDR":
            if ihdr is not None or offset != len(PNG_SIGNATURE) or length != 13:
                raise _error("INVALID_PNG", "PNG IHDR 无效")
            ihdr = data
        elif kind == b"IDAT":
            if ihdr is None:
                raise _error("INVALID_PNG", "PNG IDAT 出现在 IHDR 之前")
            idat.extend(data)
        elif kind == b"IEND":
            if length != 0:
                raise _error("INVALID_PNG", "PNG IEND 无效")
            seen_iend = True
            break
        elif kind and 65 <= kind[0] <= 90:
            raise _error("INVALID_PNG", "PNG 包含不支持的关键 chunk")
        offset = end
    if ihdr is None or not idat or not seen_iend:
        raise _error("INVALID_PNG", "PNG 缺少必要 chunk")

    width, height, depth, color_type, compression, filter_method, interlace = struct.unpack(">IIBBBBB", ihdr)
    if width <= 0 or height <= 0:
        raise _error("INVALID_DIMENSIONS", "PNG 尺寸无效")
    if depth != 8 or color_type not in (2, 6) or compression != 0 or filter_method != 0 or interlace != 0:
        raise _error("UNSUPPORTED_PIXEL_FORMAT", "仅支持非交错 8 位 RGB/RGBA PNG")
    bpp = 3 if color_type == 2 else 4
    max_pixels = ((MAX_FILE_SIZE + FIXED_HEADER_SIZE + 1024 + 2) // 3 + 1)
    if width * height > max_pixels:
        raise _error("LIMIT_EXCEEDED", "PNG 解码尺寸超过限制")
    stride = width * bpp
    expected_raw = (stride + 1) * height
    try:
        decompressor = zlib.decompressobj()
        raw = decompressor.decompress(bytes(idat), expected_raw + 1)
    except zlib.error as exc:
        raise _error("INVALID_PNG", "PNG IDAT 解压失败") from exc
    if len(raw) != expected_raw or not decompressor.eof or decompressor.unconsumed_tail or decompressor.unused_data:
        raise _error("INVALID_PNG", "PNG 解压数据长度无效")

    rows = []
    previous = bytearray(stride)
    cursor = 0
    for _ in range(height):
        filter_type = raw[cursor]
        cursor += 1
        filtered = raw[cursor:cursor + stride]
        cursor += stride
        current = bytearray(stride)
        if filter_type > 4:
            raise _error("INVALID_PNG", "PNG 使用了未知过滤器")
        for index, value in enumerate(filtered):
            left = current[index - bpp] if index >= bpp else 0
            up = previous[index]
            upper_left = previous[index - bpp] if index >= bpp else 0
            if filter_type == 0:
                predictor = 0
            elif filter_type == 1:
                predictor = left
            elif filter_type == 2:
                predictor = up
            elif filter_type == 3:
                predictor = (left + up) // 2
            else:
                predictor = _paeth(left, up, upper_left)
            current[index] = (value + predictor) & 0xFF
        rows.append(current)
        previous = current

    if color_type == 2:
        rgb = b"".join(rows)
    else:
        rgb_data = bytearray(width * height * 3)
        target = 0
        for row in rows:
            for index in range(0, len(row), 4):
                if row[index + 3] != 255:
                    raise _error("UNSUPPORTED_PIXEL_FORMAT", "RGBA PNG 包含非不透明 alpha")
                rgb_data[target:target + 3] = row[index:index + 3]
                target += 3
        rgb = bytes(rgb_data)
    return width, height, rgb

File: node-version/src/png-codec.js
const zlib = require('node:zlib');
const { crc32 } = require('./crc32');
const { FileTransferError } = require('./errors');
const { FIXED_HEADER_SIZE, MAX_FILE_SIZE } = require('./format');

const SIGNATURE = Buffer.from('89504e470d0a1a0a', 'hex');
function fail(code, message) { throw new FileTransferError(code, message); }

function chunk(kind, data) {
  const type = Buffer.from(kind, 'ascii');
  const output = Buffer.alloc(12 + data.length);
  output.writeUInt32BE(data.length, 0);
  type.copy(output, 4);
  data.copy(output, 8);
  output.writeUInt32BE(crc32(type, data), 8 + data.length);
  return output;
}

function encodeRgbPng(width, height, rgb) {
  if (width <= 0 || height <= 0 || rgb.length !== width * height * 3) fail('INVALID_DIMENSIONS', 'RGB 数据长度与图片尺寸不匹配');
  const stride = width * 3;
  const raw = Buffer.alloc((stride + 1) * height);
  for (let row = 0; row < height; row += 1) rgb.copy(raw, row * (stride + 1) + 1, row * stride, (row + 1) * stride);
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8;
  ihdr[9] = 2;
  return Buffer.concat([SIGNATURE, chunk('IHDR', ihdr), chunk('IDAT', zlib.deflateSync(raw, { level: 9 })), chunk('IEND', Buffer.alloc(0))]);
}

function paeth(a, b, c) {
  const p = a + b - c;
  const pa = Math.abs(p - a), pb = Math.abs(p - b), pc = Math.abs(p - c);
  return pa <= pb && pa <= pc ? a : pb <= pc ? b : c;
}

function decodePngToRgb(png) {
  if (!png.subarray(0, 8).equals(SIGNATURE)) fail('INVALID_PNG', '输入不是有效 PNG');
  let offset = 8;
  let ihdr = null;
  const idat = [];
  let seenIend = false;
  while (offset < png.length) {
    if (offset + 12 > png.length) fail('INVALID_PNG', 'PNG 数据被截断');
    const length = png.readUInt32BE(offset);
    const kind = png.subarray(offset + 4, offset + 8);
    const end = offset + 12 + length;
    if (end > png.length) fail('INVALID_PNG', 'PNG chunk 长度无效');
    const data = png.subarray(offset + 8, offset + 8 + length);
    if (png.readUInt32BE(offset + 8 + length) !== crc32(kind, data)) fail('INVALID_PNG', `PNG ${kind.toString('ascii')} chunk CRC 错误`);
    const name = kind.toString('ascii');
    if (name === 'IHDR') {
      if (ihdr || offset !== 8 || length !== 13) fail('INVALID_PNG', 'PNG IHDR 无效');
      ihdr = Buffer.from(data);
    } else if (name === 'IDAT') {
      if (!ihdr) fail('INVALID_PNG', 'PNG IDAT 出现在 IHDR 之前');
      idat.push(data);
    } else if (name === 'IEND') {
      if (length !== 0) fail('INVALID_PNG', 'PNG IEND 无效');
      seenIend = true;
      break;
    } else if (kind[0] >= 65 && kind[0] <= 90) fail('INVALID_PNG', 'PNG 包含不支持的关键 chunk');
    offset = end;
  }
  if (!ihdr || !idat.length || !seenIend) fail('INVALID_PNG', 'PNG 缺少必要 chunk');
  const width = ihdr.readUInt32BE(0), height = ihdr.readUInt32BE(4);
  const depth = ihdr[8], colorType = ihdr[9], compression = ihdr[10], filterMethod = ihdr[11], interlace = ihdr[12];
  if (!width || !height) fail('INVALID_DIMENSIONS', 'PNG 尺寸无效');
  if (depth !== 8 || ![2, 6].includes(colorType) || compression !== 0 || filterMethod !== 0 || interlace !== 0) fail('UNSUPPORTED_PIXEL_FORMAT', '仅支持非交错 8 位 RGB/RGBA PNG');
  const bpp = colorType === 2 ? 3 : 4;
  const maxPixels = Math.floor((MAX_FILE_SIZE + FIXED_HEADER_SIZE + 1024 + 2) / 3) + 1;
  if (width * height > maxPixels) fail('LIMIT_EXCEEDED', 'PNG 解码尺寸超过限制');
  const stride = width * bpp;
  const expectedRaw = (stride + 1) * height;
  let raw;
  try { raw = zlib.inflateSync(Buffer.concat(idat), { maxOutputLength: expectedRaw + 1 }); }
  catch { fail('INVALID_PNG', 'PNG IDAT 解压失败'); }
  if (raw.length !== expectedRaw) fail('INVALID_PNG', 'PNG 解压数据长度无效');
  const samples = Buffer.alloc(stride * height);
  let source = 0;
  for (let row = 0; row < height; row += 1) {
    const filterType = raw[source++];
    if (filterType > 4) fail('INVALID_PNG', 'PNG 使用了未知过滤器');
    const rowStart = row * stride;
    const previousStart = (row - 1) * stride;
    for (let column = 0; column < stride; column += 1) {
      const value = raw[source++];
      const left = column >= bpp ? samples[rowStart + column - bpp] : 0;
      const up = row > 0 ? samples[previousStart + column] : 0;
      const upperLeft = row > 0 && column >= bpp ? samples[previousStart + column - bpp] : 0;
      let predictor = 0;
      if (filterType === 1) predictor = left;
      else if (filterType === 2) predictor = up;
      else if (filterType === 3) predictor = Math.floor((left + up) / 2);
      else if (filterType === 4) predictor = paeth(left, up, upperLeft);
      samples[rowStart + column] = (value + predictor) & 0xff;
    }
  }
  if (colorType === 2) return { width, height, rgb: samples };
  const rgb = Buffer.alloc(width * height * 3);
  let target = 0;
  for (let index = 0; index < samples.length; index += 4) {
    if (samples[index + 3] !== 255) fail('UNSUPPORTED_PIXEL_FORMAT', 'RGBA PNG 包含非不透明 alpha');
    rgb[target++] = samples[index]; rgb[target++] = samples[index + 1]; rgb[target++] = samples[index + 2];
  }
  return { width, height, rgb };
}

module.exports = { encodeRgbPng, decodePngToRgb };
